smack-core/src/main/java/org/jivesoftware/smack/c2s/ModularXmppClientToServerConnection.java

```diff
@@ -560,10 +562,25 @@ protected void waitForFeaturesReceived(String waitFor)
     protected void newStreamOpenWaitForFeaturesSequence(String waitFor) throws InterruptedException,
                     SmackException, XMPPException {
         prepareToWaitForFeaturesReceived();
-        sendStreamOpen();
+
+        // Create StreamOpen from StreamOpenAndCloseFactory via underlying transport.
+        StreamOpenAndCloseFactory streamOpenAndCloseFactory = activeTransport.getStreamOpenAndCloseFactory();
+        CharSequence from = null;
+        CharSequence localpart = connectionInternal.connection.getConfiguration().getUsername();
+        if (localpart != null) {
+            from = XmppStringUtils.completeJidFrom(localpart, getXMPPServiceDomain());
```

Comment:
I'm not sure if this behaviour is correct in all cases, as the username can be different from the localpart of the jid.
However, I certainly wouldn't know how to fix this instead :D

Reply:
Servers will ignore the 'from' in c2s anyway, here it is merely a debug aid.